### PR TITLE
Peer testing, part I: select tester and testee

### PR DESCRIPTION
**Pull request by msgmaxim**
- keeping track of blockchain height and the latest hash
- select tester and testee on every block

### CODE REVIEW

File: httpserver/service_node.cpp [hunk 1]
```diff
@@ -316,6 +345,58 @@ void ServiceNode::attach_signature(
...
+
+    // 1. Select the tester/testee pair
+
+    std::vector<sn_record_t> members_pk = swarm_->other_nodes();
+
+    members_pk.push_back(our_address_);
+
+    std::sort(members_pk.begin(), members_pk.end());
+
+    std::cout << "peers: " << std::endl;
```
Comment (sachaaaaa) [inline httpserver/service_node.cpp]: Probably best using the filterable `BOOST_LOG_TRIVIAL` interface?

File: utils/src/utils.cpp [hunk 2]
```diff
@@ -89,4 +88,13 @@ bool parseTTL(const std::string& ttlString, uint64_t& ttl) {
     return true;
 }
 
+/// Returns a random number from [0, n); (copied from lokid)
+uint64_t uniform_distribution_portable(std::mt19937_64& mersenne_twister, uint64_t n)
+{
+  uint64_t secureMax = mersenne_twister.max() - mersenne_twister.max() % n;
```
Comment (sachaaaaa) [inline utils/src/utils.cpp]:
```suggestion
  const uint64_t secureMax = mersenne_twister.max() - mersenne_twister.max() % n;
```

File: httpserver/service_node.cpp [hunk 3]
```diff
@@ -255,13 +261,34 @@ void ServiceNode::save_bulk(const std::vector<Item>& items) {
...
-void ServiceNode::on_swarm_update(const all_swarms_t& all_swarms) {
+void ServiceNode::on_swarm_update(const block_update_t& bu) {
     if (!swarm_) {
-        BOOST_LOG_TRIVIAL(trace) << "initialized our swarm";
+        BOOST_LOG_TRIVIAL(info) << "Initialized our swarm";
         swarm_ = std::make_unique<Swarm>(our_address_);
     }
 
-    const SwarmEvents events = swarm_->update_swarms(all_swarms);
+    if (bu.height != block_height_) {
```
Comment (sachaaaaa) [inline httpserver/service_node.cpp]: I think we can be more restrictive here and only accept blocks that are newer than `block_height_`:
```cpp
if (bu.height > block_height_)
```

File: httpserver/service_node.cpp [hunk 4]
```diff
@@ -316,6 +345,58 @@ void ServiceNode::attach_signature(
...
+    uint64_t seed;
+    if (block_hash_.size() < sizeof(seed)) {
+        BOOST_LOG_TRIVIAL(error)
+            << "Could not initiate peer test: invalid block hash";
+        return;
+    }
+
+    std::memcpy(&seed, block_hash_.data(), sizeof(seed));
+    std::mt19937_64 mt(seed);
+    auto tester_idx =
```
Comment (sachaaaaa) [inline httpserver/service_node.cpp]: There are a few places that might be `const` but I'll wait for the rest of the implementation to see how they are used ;)

File: httpserver/service_node.cpp (hunk 1 above)
Comment (msgmaxim) [reply httpserver/service_node.cpp]: That's for debugging, will remove.

File: utils/src/utils.cpp (hunk 2 above)
Comment (msgmaxim) [reply utils/src/utils.cpp]: fixed

File: httpserver/service_node.cpp (hunk 3 above)
Comment (msgmaxim) [reply httpserver/service_node.cpp]: Changed to compare block hashes instead and accepting blocks lower than the current height (to allow reorgs?).

**Pull request merged**